msgraph-developer-proxy-plugins/RequestLogs/ExecutionSummaryPlugin.cs

```diff
@@ -0,0 +1,390 @@
+
+    var sortedMethodAndUrls = data.Keys.OrderBy(k => k);
+    foreach (var methodAndUrl in sortedMethodAndUrls)
+    {
+      report.AddRange(new[] {
+        "",
+        $"### {methodAndUrl}",
+      });
+
+      var sortedMessageTypes = data[methodAndUrl].Keys.OrderBy(k => k);
+      foreach (var messageType in sortedMessageTypes)
+      {
+        report.AddRange(new [] {
+          "",
+          $"#### {messageType}",
+          ""
+        });
+
+        var sortedMessages = data[methodAndUrl][messageType].Keys.OrderBy(k => k);
+        foreach (var message in sortedMessages)
+        {
+          report.Add($"- ({data[methodAndUrl][messageType][message]}) {message}");
+        }
+      }
+    }
+
+    report.AddRange(GetSummary(requestLogs));
+
+    return report.ToArray();
+  }
+
+  private string[] GetGroupedByMessageTypeReport(IEnumerable<RequestLog> requestLogs)
+  {
+    var report = new List<string>();
+    report.AddRange(GetReportTitle());
+    report.Add("## Message types");
+
+    var data = GetGroupedByMessageTypeData(requestLogs);
+
+    var sortedMessageTypes = data.Keys.OrderBy(k => k);
+    foreach (var messageType in sortedMessageTypes)
+    {
+      report.AddRange(new[] {
+        "",
+        $"### {messageType}"
+      });
+
+      if (messageType == _requestsInterceptedMessage ||
+          messageType == _requestsPassedThroughMessage)
+      {
+        report.Add("");
+
+        var sortedMethodAndUrls = data[messageType][messageType].Keys.OrderBy(k => k);
+        foreach (var methodAndUrl in sortedMethodAndUrls)
+        {
+          report.Add($"- ({data[messageType][messageType][methodAndUrl]}) {methodAndUrl}");
+        }
+      }
+      else
+      {
+        var sortedMessages = data[messageType].Keys.OrderBy(k => k);
+        foreach (var message in sortedMessages)
+        {
+          report.AddRange(new[] {
+            "",
+            $"#### {message}",
+            ""
+          });
+
+          var sortedMethodAndUrls = data[messageType][message].Keys.OrderBy(k => k);
+          foreach (var methodAndUrl in sortedMethodAndUrls)
+          {
+            report.Add($"- ({data[messageType][message][methodAndUrl]}) {methodAndUrl}");
+          }
+        }
+      }
+    }
+
+    report.AddRange(GetSummary(requestLogs));
+
+    return report.ToArray();
+  }
+
+  private string[] GetReportTitle()
+  {
+    return new string[]
+    {
+      "# Microsoft Graph Developer Proxy execution summary",
+      "",
+      $"Date: {DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss")}",
+      ""
+    };
+  }
+
+  // in this method we're producing the follow data structure
+  // request > message type > (count) message
+  private Dictionary<string, Dictionary<string, Dictionary<string, int>>> GetGroupedByUrlData(IEnumerable<RequestLog> requestLogs)
+  {
+    var data = new Dictionary<string, Dictionary<string, Dictionary<string, int>>>();
+
+    foreach (var log in requestLogs)
```

Comment:
I used a single foreach loop instead of linq so that we iterate through the data collection only once. Not to mention, it's easier to debug in the future if need be as well.

Reply:
The linq for that would likely make my head explode 🤯